### ISSUE TITLE
Inefficient Register Allocation

### ISSUE THREAD

**Issue by DavePearce** [Consensys/corset#183]
Consider an example like this:

```lisp 
(module test)
(defcolumns (P :binary@prove))
(defperspective p1 P ((A :i8)))
(defperspective p2 P ((B :i16)))
```

We can see the register allocation performed by Corset:

```
   0             test.P      𝟙   1                                       r0/test.Pι1
   1             test.p1/A   𝟠   1                                       r1/test.Aι1
   2             test.p2/B  i8   1                                       r2/test.Bι1
```

The key is that it will not allocate `A` and `B` to the same underlying column (i.e. register) because they nominally have different types.  In fact, it is safe to do this because they are not being proven.  However, there are flow on effects of allowing them to be allocated together: trace files need to use the maximum width of any column allocated to a register; the inspector will have to do the same.

Overall, its not clear to me whether or not this optimisation will add any significant value.  But it is worth considering at least.  Note also that `i8@prove` and `byte@prove` will not be allocated together either.

**Comment by DavePearce** [comment]
Another case where columns could be allocated together, but are not is e.g. this:

```
(module test)
(defcolumns (P :binary@prove))
(defperspective p1 P ((A :binary@prove)))
(defperspective p2 P ((B :binary)))
```

**Comment by delehef** [comment]
> Overall, its not clear to me whether or not this optimisation will add any significant value.

That was a design choice that was supposed to help with value validity checking while importing traces. When debugging safeguards will become less important than traces size, it will be the time to implement your proposition.

**Comment by DavePearce** [comment]
> That was a design choice that was supposed to help with value validity checking while importing traces. 

Yeah, overall now that I've dug through it and understand what's going on I think its working pretty well.  I'm not sure we'll need much more optimisation.  Its seems to be a lot of `binary` fields in particular which need to be allocated together, and as it currently works they will be!
